k2/csrc/cuda/debug.h

```diff
@@ -0,0 +1,382 @@
+ *
+ * @details
+ *  If api return status is not cudaSuccess, print and exit.
+ *  Thus it's actually an alias of K2_CUDA_CHECK_ERROR.
+ *
+ * @note
+ *  Kernel launches do not return any error code, thus checking should after it.
+ *  To wait kernel to finish, cudaDeviceSynchronize is called between.
+ *  And to avoid pre-launch error disturbing, check error before kernel launch
+ *  is also needed.
+ *  (But not necessary if programmer promise each error is take cared properly.)
+ *
+ * @remark
+ *  macro `__VA_ARGS__` is used to pass the kernel<<<...>>> as one argument,
+ *  otherwise the compiler raises a error "passed 2 arguments, but takes just 1".
+ *
+ * @code{.cpp}
+ * K2_CUDA_KERNEL_SAFE_CALL(kernel_func<<<...>>>());
+ * @endcode
+ */
+#ifndef K2_CUDA_KERNEL_SAFE_CALL
+  #ifdef K2_MAKE_ERROR_CHECK
+    #define K2_CUDA_KERNEL_SAFE_CALL(...)               \
+      do {                                              \
+        (__VA_ARGS__);                                  \
+        K2_CUDA_API_SAFE_CALL(cudaDeviceSynchronize()); \
+        K2_CUDA_CHECK_ERROR(cudaGetLastError());        \
+      } while (0)
+  #else
+    #define K2_CUDA_KERNEL_SAFE_CALL(...) (__VA_ARGS__)
+  #endif
+#endif
+
+/**
+ * @def K2_DLOG
+ *
+ * @brief Log macro for printf statements.
+ *
+ * @details
+ *  `printf` is supported by both host and device. This Log is for debugging,
+ *  the error msg is printed to the stderr. The log msg always get printed,
+ *  regardless of macro `K2_MAKE_ERROR_CHECK`. Thus it should only be
+ *  make used of in debugging.
+ *
+ * @note
+ *  This code is refered from:
+ *  https://github.com/NVlabs/cub/blob/ \
+ *  c3cceac115c072fb63df1836ff46d8c60d9eb304/cub/util_debug.cuh#L109
+ *
+ * @code{.cpp}
+ * K2_DLOG("Value is %d, string is %s ..", i, str);
+ * @endcode
+ */
+#if !defined(K2_DLOG)
+  #if !(defined(__clang__) && defined(__CUDA__))
+    #if (K2_PTX_ARCH == 0)
+      #define K2_DLOG(format, ...) printf(format,__VA_ARGS__)
+    #elif (K2_PTX_ARCH >= 200)
+      #define K2_DLOG(format, ...)                                            \
+        printf("[block (%d,%d,%d), thread (%d,%d,%d)]: " format, blockIdx.x,  \
+               blockIdx.y, blockIdx.z, threadIdx.x, threadIdx.y, threadIdx.z, \
+               __VA_ARGS__)
+    #endif
+  #else
+    /**
+     * A hack to implement the variadic printf for clang,
+     * and sielence the warning
+     */
+    #pragma clang diagnostic push
+    #pragma clang diagnostic ignored "-Wc++11-extensions"
+    #pragma clang diagnostic ignored "-Wunnamed-type-template-args"
+      template <class... Args>
+      inline __host__ __device__
+      void va_printf(char const* format, Args const&... args) {
+    #ifdef __CUDA_ARCH__
+        printf(format, blockIdx.x, blockIdx.y, blockIdx.z,
+               threadIdx.x, threadIdx.y, threadIdx.z, args...);
+    #else
+        printf(format, args...);
+    #endif
+      }
+    #ifndef __CUDA_ARCH__
```

Comment:
As you use `__CUDA_ARCH__` here directly, I really don't understand why you define
`K2_PTX_ARCH`. Could you please explain?

Reply:
Agreed, I think it's better to use __CUDA_ARCH__ instead- simpler and  more standard.

Reply:
The __CUDA_ARCH_ is not 0 for the host side, it's undefined.

Reply:
see my other comment, can do if defined(x) && x > 4

Reply:
Done